src/map-controls.js

```diff
@@ -0,0 +1,373 @@
+        // Gradually add until we hit max pitch
+        pitch = startPitch + yScale * (MAX_PITCH - startPitch);
+      }
+    }
+
+    // console.debug(startPitch, pitch);
+    return {
+      pitch: Math.max(Math.min(pitch, MAX_PITCH), 0),
+      bearing
+    };
+  }
+
+  @autobind _onTouchStart(opts) {
+    this._onMouseDown(opts);
+  }
+
+  @autobind _onTouchDrag(opts) {
+    this._onMouseDrag(opts);
+  }
+
+  @autobind _onTouchRotate(opts) {
+    this._onMouseRotate(opts);
+  }
+
+  @autobind _onTouchEnd(opts) {
+    this._onMouseUp(opts);
+  }
+
+  @autobind _onTouchTap(opts) {
+    this._onMouseClick(opts);
+  }
+
+  @autobind _onMouseDown({pos}) {
+    this._updateViewport({
+      isDragging: true,
+      startDragLngLat: this.props.unproject(pos),
+      startBearing: this.props.bearing,
+      startPitch: this.props.pitch
+    });
+  }
+
+  @autobind _onMouseDrag({pos}) {
+    if (!this.props.onChangeViewport) {
+      return;
+    }
+
+    const {startDragLngLat} = this.state;
+
+    // take the start lnglat and put it where the mouse is down.
+    assert(startDragLngLat, '`startDragLngLat` prop is required ' +
+      'for mouse drag behavior to calculate where to position the map.');
+
+    const [longitude, latitude] = this._calculateNewLngLat({
+      startDragLngLat,
+      pos
+    });
+
+    this._updateViewport({
+      longitude,
+      latitude,
+      isDragging: true
+    });
+  }
+
+  @autobind _onMouseRotate({pos, startPos}) {
+    if (!this.props.onChangeViewport || !this.props.perspectiveEnabled) {
+      return;
+    }
+
+    const {startBearing, startPitch} = this.props;
+    assert(typeof startBearing === 'number',
+      '`startBearing` prop is required for mouse rotate behavior');
+    assert(typeof startPitch === 'number',
+      '`startPitch` prop is required for mouse rotate behavior');
+
+    const {pitch, bearing} = this._calculateNewPitchAndBearing({
+      pos,
+      startPos,
+      startBearing,
+      startPitch
+    });
+
+    this._updateViewport({
+      bearing,
+      pitch,
+      isDragging: true
+    });
+  }
+
+  @autobind _onMouseUp(opt) {
+    this._updateViewport({
+      isDragging: false,
+      startDragLngLat: null,
+      startBearing: null,
+      startPitch: null
+    });
+  }
+
+  @autobind _onZoom({pos, scale}) {
+    this._updateViewport({
+      zoom: this._calculateNewZoom({relativeScale: scale}),
+      isDragging: true
+    });
+  }
+
+  @autobind _onZoomEnd() {
+    this._updateViewport({isDragging: false});
+  }
+
+  // Note that mouse move and click are handled directly by static-map
+  // onMouseMove={this._onMouseMove}
+  // onMouseClick={this._onMouseClick}
+
+  render() {
+    const {className, width, height, style} = this.props;
+    const mapEventLayerStyle = {
+      ...style,
+      width,
+      height,
+      position: 'relative',
+      cursor: this._getCursor()
+    };
+
+    return (
+      <div className={className} style={mapEventLayerStyle}>
```

Comment:
Is it necessary to render this `<div>` here or could we possibly just return `<EventManager>`?